addons/mail/controllers/attachment.py

```diff
@@ -44,9 +44,19 @@ def _make_zip(self, name, attachments):
         ]
         return request.make_response(content, headers)
 
+    def _should_generate_attachment_access_token(self, thread):
+        """Determine whether an access token should be generated when uploading
+        an attachment to the given thread.
+        """
+        # Only generate the access token if absolutely necessary (= not for internal user).
+        return request.env.user.share
+
     @http.route("/mail/attachment/upload", methods=["POST"], type="http", auth="public")
     @add_guest_to_context
     def mail_attachment_upload(self, ufile, thread_id, thread_model, is_pending=False, **kwargs):
+        thread = request.env[thread_model]._get_from_context_or_raise(int(thread_id))
+        if thread_model == "discuss.channel" and not thread.allow_public_upload and not request.env.user._is_internal():
```

Comment:
Didn't we also talk at some point about preventing it if the livechat is closed?

Reply:
I believe we could go a bit further and prevent any guest from accessing a closed live chat channel. I added an override of `discuss.channel.member@_get_as_sudo_from_context`. What do you think? 

```python
    @api.model
    def _get_as_sudo_from_context(self, channel_id):
        channel_member = super()._get_as_sudo_from_context(channel_id=channel_id)
        channel = channel_member.channel_id
        if channel_member.guest_id and channel.channel_type == "livechat" and not channel.livechat_active:
            return self.env["discuss.channel.member"].sudo()
        return channel_member
        ```

Reply:
Isn't this solution a little bit too broad?
We only want to prevent uploading attachment (and potentially posting messages), but making as if the guest doesn't exist will make issues to access anything such as potentially history, etc.

Also for this purpose I think guests and portal should be considered alike.